pkg/issuer/venafi/client/instrumentedvenaficlient.go

```diff
@@ -0,0 +1,89 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"time"
+
+	"github.com/Venafi/vcert/v4/pkg/certificate"
+	"github.com/Venafi/vcert/v4/pkg/endpoint"
+	"github.com/go-logr/logr"
+
+	logf "github.com/cert-manager/cert-manager/pkg/logs"
+	"github.com/cert-manager/cert-manager/pkg/metrics"
+)
+
+type instrumentedConnector struct {
+	conn    connector
+	metrics *metrics.Metrics
+	logger  *logr.Logger
+}
+
+var _ connector = instrumentedConnector{}
+
+func newInstumentedConnector(conn connector, metrics *metrics.Metrics, log logr.Logger) connector {
+	return instrumentedConnector{
+		conn:    conn,
+		metrics: metrics,
+		logger:  &log,
+	}
+}
+
+func (ic instrumentedConnector) ReadZoneConfiguration() (*endpoint.ZoneConfiguration, error) {
+	start := time.Now()
+	ic.logger.V(logf.TraceLevel).Info("calling ReadZoneConfiguration")
+	config, err := ic.conn.ReadZoneConfiguration()
+	labels := []string{"read_zone_configuration"}
+	ic.metrics.ObserveVenafiRequestDuration(time.Since(start), labels...)
+	return config, err
+}
+
+func (ic instrumentedConnector) RequestCertificate(req *certificate.Request) (string, error) {
+	start := time.Now()
+	ic.logger.V(logf.TraceLevel).Info("calling RequestCertificate")
+	reqID, err := ic.conn.RequestCertificate(req)
+	labels := []string{"request_certificate"}
+	ic.metrics.ObserveVenafiRequestDuration(time.Since(start), labels...)
+	return reqID, err
+}
+
+func (ic instrumentedConnector) RetrieveCertificate(req *certificate.Request) (*certificate.PEMCollection, error) {
+	start := time.Now()
+	ic.logger.V(logf.TraceLevel).Info("calling RetrieveCertificate")
+	pemCollection, err := ic.conn.RetrieveCertificate(req)
+	labels := []string{"retrieve_certificate"}
+	ic.metrics.ObserveVenafiRequestDuration(time.Since(start), labels...)
+	return pemCollection, err
+}
+
+func (ic instrumentedConnector) Ping() error {
+	start := time.Now()
+	ic.logger.V(logf.TraceLevel).Info("calling Ping")
+	err := ic.conn.Ping()
+	labels := []string{"ping"}
+	ic.metrics.ObserveVenafiRequestDuration(time.Since(start), labels...)
+	return err
+}
+
+func (ic instrumentedConnector) RenewCertificate(req *certificate.RenewalRequest) (string, error) {
+	start := time.Now()
+	ic.logger.V(logf.TraceLevel).Info("calling RenewCertificate")
+	reqID, err := ic.conn.RenewCertificate(req)
+	labels := []string{"renew_certificate"}
+	ic.metrics.ObserveVenafiRequestDuration(time.Since(start), labels...)
+	return reqID, err
+}
```

Comment:
It's a shame there isn't some generic way to wrap and time a given function. 
If we could create a generic wrapper we could unit-test that.

Reply:
BTW I'm not suggesting that you need to do this here, just seems like there must be a less repetitive way of getting metrics for any given interface. 

I wonder how easy it would be to write a go generator to produce this kind of wrapper for any interface?

 Or a wrapper around the HTTP client API which could be passed to the vcert client. E.g.
 * https://github.com/travelaudience/go-promhttp#wrapping-the-httpclient

Reply:
Thanks @wallrj I think having a more generic way to obtain metrics for external API call latencies would be useful (also logging of call latencies could be added to this).
I have created a new issue for that for now https://github.com/cert-manager/cert-manager/issues/5075
Perhaps it'd make sense to do that once we have a third such API whose call latencies we want to measure ([rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) - so as to make a better abstraction?)